zeusd/src/devices/cpu/mod.rs

```diff
@@ -0,0 +1,170 @@
+// RAPL CPU
+// Real RAPL interface.
+#[cfg(target_os = "linux")]
+mod linux;
+#[cfg(target_os = "linux")]
+pub use linux::RaplCpu;
+
+// Fake Rapl interface for dev and testing on macOS.
+#[cfg(target_os = "macos")]
+mod macos;
+#[cfg(target_os = "macos")]
+pub use macos::RaplCpu;
+
+use serde::Serialize;
+use std::path::PathBuf;
+use std::sync::RwLock;
+use std::time::Instant;
+use tokio::sync::mpsc::{Sender, UnboundedReceiver, UnboundedSender};
+use tracing::Span;
+
+use crate::error::ZeusdError;
+
+pub struct PackageInfo {
+    pub index: u32,
+    pub name: String,
+    pub energy_uj_path: PathBuf,
+    pub max_energy_uj: u64,
+
+    num_wraparounds: RwLock<u64>,
+}
+
+#[derive(Serialize)]
+pub struct RaplResponse {
+    cpu_energy_uj: Option<u64>,
+    dram_energy_uj: Option<u64>,
+}
+
+pub trait CpuManager {
+    fn device_count() -> Result<u32, ZeusdError>;
+    fn get_available_fields(index: u32) -> Result<(PackageInfo, Option<PackageInfo>), ZeusdError>;
+    fn get_cpu_energy(&self) -> Result<u64, ZeusdError>;
+    fn get_dram_energy(&self) -> Result<Option<u64>, ZeusdError>;
+    fn stop_monitoring(&mut self);
+}
+
+pub type CpuCommandRequest = (
+    CpuCommand,
+    Option<Sender<Result<RaplResponse, ZeusdError>>>,
+    Instant,
+    Span,
+);
+
+#[derive(Clone)]
+pub struct CpuManagementTasks {
+    // Senders to the CPU management tasks. index is the CPU ID.
+    senders: Vec<UnboundedSender<CpuCommandRequest>>,
+}
+
+impl CpuManagementTasks {
+    pub fn start<T>(cpus: Vec<T>) -> anyhow::Result<Self>
```

Comment:
A library method returning `anyhow::Result` is a little bit weird. As you know, Rust convention is to create an internal library/crate (`zeusd` here) and then the application binary (`main.rs`) consumes that library to build the application. I think Anyhow is for the top-level application (`main.rs`) that consumes the `zeusd` crate, and the crate itself returns built-in `Result`s that possibly contain errors created with `thiserror` (i.e., `ZeusdError`).

Reply:
I think we could change so it is:
```
    pub fn start<T>(cpus: Vec<T>) -> Result<Self, ZeusdError>
```
Should I do the same for `start` in `GpuManagementTasks`?

Reply:
Yes please